components/calendar/src/calendar_arithmetic.rs

```diff
@@ -34,41 +34,48 @@ impl<C: CalendarArithmetic> ArithmeticDate<C> {
             marker: PhantomData,
         }
     }
+
     #[inline]
-    pub fn offset_date(&mut self, mut offset: DateDuration<C>) {
-        self.year += offset.years;
-        self.month += offset.months as u8;
-        offset.months = 0;
-
-        offset.days += offset.weeks * 7;
-
-        offset.days += self.day as i32 - 1;
-        self.day = 1;
-
-        while offset.days != 0 {
-            if offset.days < 0 {
-                self.month -= 1;
-                let month_days = C::month_days(self.year, self.month);
-                if (-offset.days) > month_days as i32 {
-                    offset.days += month_days as i32;
-                } else {
-                    self.day = 1 + (month_days as i8 + offset.days as i8) as u8;
-                    offset.days = 0;
-                }
+    pub fn offset_days(&mut self, mut day_offset: i32) {
+        while day_offset != 0 {
+            let month_days = C::month_days(self.year, self.month);
+            if self.day as i32 + day_offset > month_days as i32 {
+                self.offset_months(1);
+                day_offset -= month_days as i32;
+            } else if self.day as i32 + day_offset < 1 {
+                self.offset_months(-1);
```

Comment:
issue: There's an edge case here where `self.day` is a valid day for the current month but is not a valid day for the previous month.

Since 1 is always a valid day, when going both forward _and_ backwards, we should `offset_months` and _then_ reset `self.day` such that it is 1.

Reply:
Actually, that might involve other problems in offset_months.

An alternate route is to have a "fixup" step in `offset_date()` that ensures that after offset_days we are still in bounds forthe current month; and if not, sets days to 1, offset_days to days, and re-calls offset_days

Reply:
One error case for this would be, for example, adding 1 month and 0 days to March 31

I think ideally offset_days should offset to Day 1 of the previous/next month wherever possible, and offset_date should be able to fixup any out of bounds things from offset_months

Reply:
Done - Agreed that it's cleaner to fix from `offset_date()`. I also made `offset_months` and `offset_days` non-public so they can only be used as helpers inside `offset_date()`, which should always produce in-bound results.